### PR TITLE
Allow windows to share browsing contexts.

### LINKED ISSUE #13608: Transplant WindowProxy objects
As part of #633 we should use `JS_TransplantObject` to transplant window proxy objects when an iframe navigates. At the moment, if a same-origin iframe stores its window in its parent, then same-origin navigates, the window object in the new document is different from the old one.

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR allows different `Window` objects in the same browsing context to share a `BrowsingContext` object.

SpiderMonkey requires a `WindowProxy` object to be in the same compartment as its `Window`, so when a `WindowProxy` changes `Window`, we have to brain-transplant it. In turn this requires the reflector of a `BrowsingContext` to be mutable.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13608 and #14843 
- [X] These changes do not require tests because an existing test catches this (`/html/browsers/the-window-object/Window-document.html` is now `PASS`)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15120)

### CODE REVIEW

**Comment by jdm** [comment]
I think my changes in #15118 may make some of the mutable Reflector changes unnecessary (particularly if we modify rust-mozjs to make Heap::set not require &mut self).

**Comment by asajeffrey** [comment]
You're right, we could use `reflector.rootable()` to set the reflector.

**Comment by asajeffrey** [comment]
@jdm: I used `reflector.rootable()` to set the reflector, rather than having the reflector inside a `RefCell`.

**Comment by asajeffrey** [comment]
@cbrewster does this PR fix the problems you were having with webstorage tests? http://logs.glob.uno/?c=mozilla%23servo&s=11+Jan+2017&e=11+Jan+2017#c591494

**Comment by asajeffrey** [comment]
For the record, here's the IRC conversation I had with @bzbarsky about brain transplanting window proxy objects: http://logs.glob.uno/?c=mozilla%23jsapi&s=19+Jan+2017&e=19+Jan+2017#c819204

File: components/script/dom/browsingcontext.rs [hunk 1]
```diff
@@ -86,6 +100,50 @@ impl BrowsingContext {
...
+            rooted!(in(cx) let new_window_proxy = JS_TransplantObject(cx, old_window_proxy, new_window_proxy.handle()));
+            debug!("Transplanted window proxy is {:p}.", new_window_proxy.get());
+
+            // Transfer ownership of this browsing context from the old window proxy to the new one.
+            SetProxyExtra(new_window_proxy.get(), 0, &PrivateValue(self as *const _ as *const _));
+
+            // Set the window proxy.
+            SetWindowProxy(cx, window_jsobject, new_window_proxy.handle());
+
+            // Create a new reflector.
```
Comment (jdm) [inline components/script/dom/browsingcontext.rs]: `Update the reflector.`

File: components/script/dom/browsingcontext.rs [hunk 2]
```diff
@@ -86,6 +100,50 @@ impl BrowsingContext {
...
+            // pointing to the new window proxy.
+            rooted!(in(cx) let new_window_proxy = NewWindowProxy(cx, window_jsobject, handler));
+            debug!("Transplanting window proxy from {:p} to {:p}.", old_window_proxy.get(), new_window_proxy.get());
+            rooted!(in(cx) let new_window_proxy = JS_TransplantObject(cx, old_window_proxy, new_window_proxy.handle()));
+            debug!("Transplanted window proxy is {:p}.", new_window_proxy.get());
+
+            // Transfer ownership of this browsing context from the old window proxy to the new one.
+            SetProxyExtra(new_window_proxy.get(), 0, &PrivateValue(self as *const _ as *const _));
+
+            // Set the window proxy.
```
Comment (jdm) [inline components/script/dom/browsingcontext.rs]: This comment is redundant given the function name. How about `Notify the JS engine that the window proxy has changed.`?

File: components/script/dom/document.rs [hunk 3]
```diff
@@ -184,12 +184,12 @@ pub struct Document {
     node: Node,
     window: JS<Window>,
     /// https://html.spec.whatwg.org/multipage/#concept-document-bc
```
Comment (jdm) [inline components/script/dom/document.rs]: This comment doesn't belong any more.

File: components/script/dom/document.rs [hunk 4]
```diff
@@ -1998,7 +2002,7 @@ impl Document {
     }
 
     pub fn new(window: &Window,
-               browsing_context: Option<&BrowsingContext>,
+               has_browsing_context: bool,
```
Comment (jdm) [inline components/script/dom/document.rs]: I lean towards a self-documenting enum here, given that every caller passes a boolean literal.

File: components/script/script_thread.rs [hunk 5]
```diff
@@ -392,13 +392,15 @@ impl<'a> Iterator for DocumentsIter<'a> {
...
 
-
 #[derive(JSTraceable)]
 // ScriptThread instances are rooted on creation, so this is okay
 #[allow(unrooted_must_root)]
 pub struct ScriptThread {
     /// The documents for pipelines managed by this thread
     documents: DOMRefCell<Documents>,
+    /// The browsing contexts known by this thread
+    /// TODO: this map grows, but never shrinks.
```
Comment (jdm) [inline components/script/script_thread.rs]: We should file an issue about this and mention it here.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/browsingcontext.rs (hunk 2 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Done

File: components/script/dom/browsingcontext.rs (hunk 1 above)
Comment (asajeffrey) [inline components/script/dom/browsingcontext.rs]: Oops. Done.

File: components/script/dom/document.rs (hunk 3 above)
Comment (asajeffrey) [inline components/script/dom/document.rs]: Oops. Done.

File: components/script/dom/document.rs (hunk 4 above)
Comment (asajeffrey) [inline components/script/dom/document.rs]: Done.

File: components/script/script_thread.rs (hunk 5 above)
Comment (asajeffrey) [inline components/script/script_thread.rs]: Done. It's #15258.

**Review by jdm — APPROVED** [review]

**Comment by asajeffrey** [comment]
Squashed. @bors-servo: r=jdm

**Comment by cbrewster** [comment]
```
error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/components/script/dom/browsingcontext.rs:143:42
    |
143 |             *self.reflector.rootable() = new_window_proxy.get();
    |                                          ^^^^^^^^^^^^^^^^^^^^^^ expected struct `js::jsapi::Heap`, found *-ptr
    |
    = note: expected type `js::jsapi::Heap<*mut js::jsapi::JSObject>`
               found type `*mut js::jsapi::JSObject`
```

**Comment by asajeffrey** [comment]
Oh rats, `reflector.rootable()` now returns a `&Heap<*mut JSObject>` rather than a `*mut *mut JSObject`.

@jdm should I go back to `Reflector` being a `RefCell`? Or is there a better way to update the reflector?

**Comment by jdm** [comment]
It's not clear to me why Heap::set requires `&mut self`. Without that restriction you wouldn't have any problems.

**Comment by jdm** [comment]
I propose making Heap::set take &self. It stores an UnsafeCell internally, and I think it was just following the pattern of the other mutable types in https://github.com/servo/rust-mozjs/blob/master/src/rust.rs.

**Comment by asajeffrey** [comment]
OK,. Slightly annoyingly this means a PR to rust-mozjs, but it's probably the easiest solution.

**Comment by asajeffrey** [comment]
Filed https://github.com/servo/rust-mozjs/pull/328.

**Comment by asajeffrey** [comment]
That would be caused by it picking up the old version of rust-mozjs. Wait for https://github.com/servo/rust-mozjs/pull/329 to land...

**Comment by asajeffrey** [comment]
https://github.com/servo/rust-mozjs/pull/329 has landed. Rebased. @bors-servo: r=jdm

**Pull request merged**